### ISSUE TITLE
Cannot choose carrier on checkout page

### ISSUE THREAD

**Issue by dosbiner** [thirtybees/thirtybees#550]
hi, after upgrade to 1.0.5 I can't choose carrier on checkout page. I make gif for showing it https://prnt.sc/k4ua56

debug mode on showing this error:
Notice: Undefined variable: delivery_option in /home/dev/webapps/dev11/classes/Cart.php on line 1039

I replace Cart.php from 1.0.4 and then the problem gone, but I think its just temporary solution for me.

I've tried both on production site and fresh tb 1.0.5.
Nginx
PHP 7.2.5
MariaDB

**Comment by Traumflug** [comment]
I think I caught it: https://github.com/thirtybees/thirtybees/commit/05997b6feea465031d639fe1ac006105298d03f8

Would you be kind enough to test this version of Cart.php? https://raw.githubusercontent.com/thirtybees/thirtybees/05997b6feea465031d639fe1ac006105298d03f8/classes/Cart.php

**Comment by dosbiner** [comment]
error not showing anymore, but still cannot choose the carrier. have tried to switch php 7.0, 7.1, and switch to apache+nginx, still cannot choose the carrier

**Comment by Traumflug** [comment]
Then there's another problem :-) I'll experiment a bit myself.

**Comment by Traumflug** [comment]
Hmm. It works fine in my local installation, the radio buttons stick.

As these buttons jump back immediately, it might be some JavaScript problem. Can you enable the web console (Firefox: hit F12) and look for messages there?

**Comment by dosbiner** [comment]
there are no error logs in console, I have tried Firefox and Chrome.
I send you message on TB forum the url site, BO Access, and FTP access if you need to check the problem

**Comment by Traumflug** [comment]
Thanks for the credentials.

- Went to the front office page.
- Clicked on product Coffee.
- Clicked "Add to cart"
- Clicked "Proceed to checkout"
- Scrolled down to Delivery methods, ignoring all the login and address stuff.
- Clicked on various radio buttons, each one sticks.

Then I uploaded `classes/Cart.php` from Git tag *1.0.5*, reloaded the page and did see the selection jumping back.

Then I uploaded `classes/Cart.php` from current *1.0.x* (which matches that of branch *carrier* of yesterday), reloaded the web page and the problem no longer happened. Carriers stick again.

Conclusion: your shop works just fine.

If it still doesn't for you, I'd simply put this aside for a week. There are certainly other things to do. Internet stuff has sometimes self healing magic.

**Comment by getdatakick** [comment]
I can reproduce this issue as well. This is an injection from commit e7095a7b7ce3e0607b4d4a1795c1fc76b7f779bd

The problem is on line [1037](https://github.com/thirtybees/thirtybees/blob/371cf79367b7399a0530f1a75dda07d1c4d43340/classes/Cart.php#L1037) 
```php
json_decode($this->delivery_option);
``` 
returns object instead of array (which is expected in [code later](https://github.com/thirtybees/thirtybees/blob/371cf79367b7399a0530f1a75dda07d1c4d43340/classes/Cart.php#L1039) )

Fix is fairy simple: 
```php
json_decode($this->delivery_option, true);
```

**Comment by dosbiner** [comment]
wow you right, its working now, thank you datakick :+1: 
this issue solved now @Traumflug

**Comment by Traumflug** [comment]
Excellent!

As you can see by the commits mentioned above, I extended this finding to an audit of all `json_decode()` usages. Which resulted in quite a few edits.

Closing, as the issue is solved. If there's anything left, don't hesitate to reopen it.

**Comment by drudesign** [comment]
Having same issue with up to date Cart.php file - any ideas?

**Comment by getdatakick** [comment]
@drudesign I believe that's already fixed by cf7adb26a15f289d23a563db95086d6ad494d860

**Comment by drood89** [comment]
Hello. I have the same problem. I checked the solution:
https://github.com/thirtybees/thirtybees/commit/cf7adb26a15f289d23a563db95086d6ad494d860

Did not help. I still had:

Notice: Undefined variable: delivery_option in /home/modwork-io/web/as1.modwork.io/public_html/classes/Cart.php on line 1039

I checked the file:

https://raw.githubusercontent.com/thirtybees/thirtybees/05997b6feea465031d639fe1ac006105298d03f8/classes/Cart.php

And that solved the problem.
